src/SessionSetup/ServiceDeployManager.cpp

```diff
@@ -305,57 +306,91 @@ orbit_base::Future<ErrorMessageOr<void>> ServiceDeployManager::CopyFileToLocal(
   orbit_base::Promise<ErrorMessageOr<void>> promise;
   auto future = promise.GetFuture();
 
+  // This schedules the call of `CopyFileToLocalImpl` on the background thread.
   QMetaObject::invokeMethod(this,
                             [this, source = std::move(source), destination = std::move(destination),
                              promise = std::move(promise)]() mutable {
-                              promise.SetResult(CopyFileToLocalImpl(source, destination));
+                              CopyFileToLocalImpl(std::move(promise), source, destination);
                             });
 
   return future;
 }
 
-ErrorMessageOr<void> ServiceDeployManager::CopyFileToLocalImpl(std::string_view source,
-                                                               std::string_view destination) {
+void ServiceDeployManager::CopyFileToLocalImpl(orbit_base::Promise<ErrorMessageOr<void>> promise,
+                                               std::string_view source,
+                                               std::string_view destination) {
   CHECK(QThread::currentThread() == thread());
-  LOG("Copying remote \"%s\" to local \"%s\"", source, destination);
-
-  auto sftp_channel = StartSftpChannel();
 
-  if (!sftp_channel) {
-    return ErrorMessage(
-        absl::StrFormat(R"(Unable to start sftp channel to copy the remote "%s" to "%s": %s)",
-                        source, destination, sftp_channel.error().message()));
+  if (copy_file_operation_in_progress_) {
+    waiting_copy_operations_.emplace_back([this, promise = std::move(promise),
+                                           source = std::string{source},
+                                           destination = std::string{destination}]() mutable {
+      CopyFileToLocalImpl(std::move(promise), source, destination);
+    });
+    return;
   }
 
-  orbit_ssh_qt::SftpCopyToLocalOperation operation{&session_.value(), sftp_channel.value().get()};
+  copy_file_operation_in_progress_ = true;
 
-  orbit_qt_utils::EventLoop loop{};
-  auto quit_handler =
-      ConnectQuitHandler(&loop, &operation, &orbit_ssh_qt::SftpCopyToLocalOperation::stopped);
-  auto error_handler = ConnectErrorHandler(&loop, &operation,
-                                           &orbit_ssh_qt::SftpCopyToLocalOperation::errorOccurred);
-  auto cancel_handler = ConnectCancelHandler(&loop, this);
+  LOG("Copying remote \"%s\" to local \"%s\"", source, destination);
 
-  operation.CopyFileToLocal(source, destination);
+  // NOLINTNEXTLINE - Unfortunately we have to fall back to a raw `new` here.
+  auto operation =
+      new orbit_ssh_qt::SftpCopyToLocalOperation{&session_.value(), sftp_channel_.get()};
```

Comment:
Just out of interest / to see if I understand everything: Since `CopyFileToLocalImpl` is called on the background thread, the thread associated with `operation` is also the background thread. Hence the `Qt::connect(operation, ...)` that happen later, will also be executed on the background thread

Reply:
Yes, exactly. That's one reason I went with this solution. All in here is running on the background thread.